crates/bioauth-flow-rpc/src/lib.rs

```diff
@@ -366,14 +369,42 @@ where
             )
             .map_err(AuthenticateError::RuntimeApi).map_err(errtype)?;
 
-        self.pool
+        let mut watch = self.pool
             .submit_and_watch(
                 &sp_api::BlockId::Hash(at),
                 sp_runtime::transaction_validity::TransactionSource::Local,
                 ext,
             )
             .await
-            .map_err(AuthenticateError::BioauthTx).map_err(errtype)?;
+            .map_err(AuthenticateError::BioauthTx).map_err(errtype)?.fuse();
+
+        loop {
+            let tx_status = watch.select_next_some().await;
+
+            match tx_status {
+                TransactionStatus::Finalized(_)=> break,
+                TransactionStatus::Retracted(_) => Err(
+                    errtype(AuthenticateError::TxNotFinalized(TxNotFinalizedError::Retracted))
+                )?,
+                TransactionStatus::Usurped(_) => Err(
+                    errtype(AuthenticateError::TxNotFinalized(TxNotFinalizedError::Usurped))
+                )?,
+                TransactionStatus::Dropped => Err(
+                        errtype(AuthenticateError::TxNotFinalized(TxNotFinalizedError::Dropped))
+                    )?,
+                TransactionStatus::FinalityTimeout(_) => Err(
+                        errtype(AuthenticateError::TxNotFinalized(TxNotFinalizedError::FinalityTimeout))
+                    )?,
+                TransactionStatus::Invalid => Err(
+                        errtype(AuthenticateError::TxNotFinalized(TxNotFinalizedError::Invalid))
+                    )?,
+                // Valid and expected statuses of transaction to be finalized.
+                TransactionStatus::Ready
+                    | TransactionStatus::Broadcast(_)
+                    | TransactionStatus::InBlock(_)
+                    | TransactionStatus::Future => {},
```

Comment:
Would be nice to log the intermediate status updates here

Reply:
https://github.com/humanode-network/humanode/pull/1018/commits/012a23941b2bf6cce2d3de110d8e37983d5cd4cd